src/IO/Observer/VolumeActions.hpp

```diff
@@ -0,0 +1,163 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <unordered_map>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Tensor/TensorData.hpp"
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/File.hpp"
+#include "IO/H5/VolumeData.hpp"
+#include "IO/Observer/ArrayComponentId.hpp"
+#include "IO/Observer/ObservationId.hpp"
+#include "IO/Observer/ObserverComponent.hpp"
+#include "IO/Observer/Tags.hpp"
+#include "IO/Observer/TypeOfObservation.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Info.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace observers {
+namespace ThreadedActions {
+/// \cond
+struct WriteVolumeData;
+/// \endcond
+
+/*!
+ * \ingroup ObserverGroup
+ * \brief Send volume tensor data to the observer.
+ *
+ * \warning  Currently this action can only be called once per `observation_id`
+ * per `array_component_id`. Calling it more often is undefined behavior and may
+ * result in incomplete data being written and/or memory leaks. The reason is
+ * that we do not register the number of times a component with an ID at a
+ * specific observation id will call the observer. However, this will be a
+ * feature implemented in the future.
+ */
+struct ContributeVolumeData {
+  template <typename... DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent, size_t Dim,
+            Requires<sizeof...(DbTags) != 0> = nullptr>
+  static auto apply(db::DataBox<tmpl::list<DbTags...>>& box,
+                    tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/,
+                    const observers::ObservationId& observation_id,
+                    const observers::ArrayComponentId& array_component_id,
+                    std::vector<TensorComponent>&& in_received_tensor_data,
+                    const Index<Dim>& received_extents) noexcept {
+    db::mutate<Tags::VolumeDataLock, Tags::TensorData>(
```

Comment:
What if we prevent `db::mutate` and `db::get` from returning lock types, and add a separate function for retrieving lock types to the DataBox? This would allow you to get the lock, acquire it, mutate, release. The `get_lock` function I think then could get away with ignoring the `mutate_locked_box_` flag since the lock needs to be used separately anyway. I'll need to check how strongly typed Charm++'s lock is (It may be another of these pointer-to-void things), then we could wrap it in our own class to make the type checking possible.

Reply:
Never mind, there is a lock passed in to the threaded actions. I'm just forgetting how I wrote the code...

Reply:
Thinking about this a fair bit more (and having gotten the locks sorted out), I think what we actually want is to first collect on a group to avoid having to deal with locks, then collect the results from the groups on a nodegroup, followed by writing/further reduction/whatever we find is best. This approach will require acquiring a lock `num_elements_per_node/num_cores_per_node` fewer times, which will be between 10-1000.